### PR TITLE
Bluebird remove return, reflect, fromCallback

### PR DESCRIPTION
**Pull request by maximelkin**

### CODE REVIEW

**Comment by kibertoad** [comment]
It seems to be failing on Node 12 consistently.

**Comment by maximelkin** [comment]
Hm, strange errors, I will research this

**Comment by maximelkin** [comment]
So error got passed through try-catch, really strange

```
Error: This socket has been ended by the other party
    at Socket.writeAfterFIN [as write] (net.js:441:14)
    at Connection.write (/home/travis/build/knex/knex/node_modules/mysql2/lib/connection.js:217:17)
    at Connection.writePacket (/home/travis/build/knex/knex/node_modules/mysql2/lib/connection.js:262:12)
    at Quit.start (/home/travis/build/knex/knex/node_modules/mysql2/lib/commands/quit.js:24:16)
    at Quit.execute (/home/travis/build/knex/knex/node_modules/mysql2/lib/commands/command.js:39:22)
    at Connection.handlePacket (/home/travis/build/knex/knex/node_modules/mysql2/lib/connection.js:408:32)
    at Connection.addCommand (/home/travis/build/knex/knex/node_modules/mysql2/lib/connection.js:430:12)
    at Connection.end (/home/travis/build/knex/knex/node_modules/mysql2/lib/connection.js:796:26)
    at internal/util.js:278:30
    at new Promise (<anonymous>)
    at bound end (internal/util.js:277:12)
    at Client_MySQL2.destroyRawConnection (/home/travis/build/knex/knex/lib/dialects/mysql/index.js:85:20)
```

**Comment by maximelkin** [comment]
I misunderstood, error not get passed through try-catch  
Error because of trying to send select inside already closed connection

**Comment by maximelkin** [comment]
connection.end called after last command fail, which cleared all events, so 'error' event emitted to nowhere and emitted as unhandled error

**Comment by maximelkin** [comment]
https://github.com/nodejs/node/issues/24111

**Comment by maximelkin** [comment]
Before node12
1) connection.end called
2) also error emit in 1) synchroniously
3) catch
4) finally

Was
1) connection.end called (process.nextTick error triggered)
2) Bluebird.catch
3) error emitted
4) Bluebird.finally

Now
1) connection.end (process.nextTick error triggered)
2) catch (e) {}
3) finally {}
4) error emitted

Bluebird solution worked because of additional microtask from .catch?

**Comment by kibertoad** [comment]
@maximelkin Can we add additional listener to catch the error when it is thrown asynchronously?

**Comment by maximelkin** [comment]
@kibertoad can this trigger leaks? When to unsubscribe this listener?

**Comment by maximelkin** [comment]
e2502602359c22a5098e95267217a4fbb8655eba removeAllListeners was added

**Comment by maximelkin** [comment]
7069ce509ed1ed6b1c3a0e9dde8d27a29bfcb30d <- here removeAllListeners added

**Comment by maximelkin** [comment]
@kibertoad that do you think?

File: lib/dialects/mysql/index.js [hunk 1]
```diff
@@ -78,12 +79,13 @@ Object.assign(Client_MySQL.prototype, {
...
   // when a connection times out or the pool is shutdown.
-  destroyRawConnection(connection) {
-    return Bluebird.fromCallback(connection.end.bind(connection))
-      .catch((err) => {
-        connection.__knex__disposed = err;
-      })
-      .finally(() => connection.removeAllListeners());
+  async destroyRawConnection(connection) {
+    try {
+      const end = promisify(connection.end.bind(connection));
```
[the remark this answers is not in the document]
Comment (elhigu) [reply lib/dialects/mysql/index.js]: Sorry, yeah now I got it when I read the node-mysql docs... I would still kind of prefer

```
promisify(cb => connection.end(cb));
```

[the remark this answers is not in the document]
Comment (maximelkin) [reply lib/dialects/mysql/index.js]: done

File: lib/dialects/oracle/index.js [hunk 2]
```diff
@@ -96,7 +97,8 @@ Object.assign(Client_Oracle.prototype, {
   // Used to explicitly close a connection, called internally by the pool
   // when a connection times out or the pool is shutdown.
   destroyRawConnection(connection) {
-    return Bluebird.fromCallback(connection.close.bind(connection));
+    const close = promisify(connection.close.bind(connection));
```
[the remark this answers is not in the document]
Comment (maximelkin) [reply lib/dialects/oracle/index.js]: done

File: lib/dialects/postgres/index.js [hunk 3]
```diff
@@ -134,7 +135,8 @@ Object.assign(Client_PG.prototype, {
   // Used to explicitly close a connection, called internally by the pool
   // when a connection times out or the pool is shutdown.
   destroyRawConnection(connection) {
-    return Bluebird.fromCallback(connection.end.bind(connection));
+    const end = promisify(connection.end.bind(connection));
```
[the remark this answers is not in the document]
Comment (maximelkin) [reply lib/dialects/postgres/index.js]: done

File: lib/dialects/sqlite3/index.js [hunk 4]
```diff
@@ -78,7 +79,8 @@ Object.assign(Client_SQLite3.prototype, {
   // Used to explicitly close a connection, called internally by the pool when
   // a connection times out or the pool is shutdown.
   destroyRawConnection(connection) {
-    return Bluebird.fromCallback(connection.close.bind(connection));
+    const close = promisify(connection.close.bind(connection));
```
[the remark this answers is not in the document]
Comment (maximelkin) [reply lib/dialects/sqlite3/index.js]: done

File: lib/dialects/mysql/index.js [hunk 5]
```diff
@@ -78,12 +79,13 @@ Object.assign(Client_MySQL.prototype, {
...
+      return await end();
+    } catch (err) {
+      connection.__knex__disposed = err;
+    }
```
[the remark this answers is not in the document]
Comment (maximelkin) [reply lib/dialects/mysql/index.js]: I run knex-stress-test for a some minutes (with process.memoryUsage().heapUsed logging)
memory usage is stable, so no leak introduced

**Comment by maximelkin** [comment]
>Dropping that removeAllListeners call seems wrong.

@elhigu why?

**Comment by elhigu** [comment]
> > Dropping that removeAllListeners call seems wrong.
> 
> @elhigu why?

Because its addition didn't look like an accident and there was not proper investigation done, if it can be removed or not.

File: lib/dialects/mysql/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (elhigu) [reply lib/dialects/mysql/index.js]: Sounds promising. Did you try it with both mysql and mysql2 drivers? 

Though if that listener removal is not necessary, shouldn't this function just return end() and leave it to caller to set __knex__disposed like in every other driver 🤔 

One more thing... if that `Error: This socket has been ended by the other party` is now not leaked, I suppose that error is still happening, but it is ignored somewhere (by existence of some event handler maybe). So when it is ignored how does that get recorded to  `connection.__knex__disposed` or does it even matter...

Also I'm pretty confused how is it possible that error is leaking and where that leak is happening exactly (stack trace showed lib/dialects/mysql/index.js:85:20, but then it should have been handled by caller's error handler). And when it is leaked why caller's error handling is not catching it? That was an async function after all.

This is a part of really fundamental functionality of knex and may cause hard to catch failures for huge user base if it is not working properly.I would really like to understand the problem properly before trying to patch it.

[the remark this answers is not in the document]
Comment (elhigu) [reply lib/dialects/mysql/index.js]: Can I easily reproduce the original error to investigate it a bit more? Did it happen in some very specific test or every time?

[the remark this answers is not in the document]
Comment (maximelkin) [reply lib/dialects/mysql/index.js]: Install 12 node
Add finally block with removeAllListeners
run test "#2321 dead connections are not evicted from pool"

**Comment by maximelkin** [comment]
As temporal solution we can add delay before removeAllListeners, so it would work like before

**Comment by kibertoad** [comment]
@maximelkin I believe https://lodash.com/docs/4.17.14#defer does exactly this?

**Comment by maximelkin** [comment]
@kibertoad yes

**Comment by maximelkin** [comment]
But i still think what we should review removeAllListeners usage, maybe in separate issue

**Comment by maximelkin** [comment]
@elhigu @kibertoad done

**Comment by Atifnaquee95** [comment]
image
image
can you please help in this i am not able to get data

**Pull request merged**